### ISSUE TITLE
RepeatLayer

### ISSUE THREAD

**Issue by skaae** [Lasagne/Lasagne#362]
I think we should add a repeat layer similar to Keras' http://keras.io/layers/core/#repeatvector.

Using the slicing layer and the repeat layer its very easy to implement an encoder/decoder structure.

I have been using the following code for some RNN+transformerlayer experiments: 

```Python
class Repeat(lasagne.layers.Layer):
    def __init__(self, incoming, n, **kwargs):
        super(Repeat, self).__init__(incoming, **kwargs)
        self.n = n

    def get_output_shape_for(self, input_shape):
        return tuple([input_shape[0], self.n] + list(input_shape[1:]))

    def get_output_for(self, input, **kwargs):
        tensors = [input]*self.n
        stacked = theano.tensor.stack(*tensors)
        dim = [1, 0] + range(2, input.ndim + 1)
        return stacked.dimshuffle(dim)

```

**Comment by benanne** [comment]
Using `T.stack` for this seems pretty inefficient since that will result in a copy. Isn't there a way to do this with broadcasting or something? It seems unnecessary to copy the same thing n times.

**Comment by f0k** [comment]
> Isn't there a way to do this with broadcasting or something?

Yes, if I see correctly, what Søren does is partly equivalent to `return input.dimshuffle(0, 'x', *range(2, input.ndim))`.
The problem is that Søren sometimes explicitly needs a tensor of the given shape to be fed to `scan`, and Theano doesn't offer a functionality similar to `np.lib.stride_tricks.as_strided` -- that would allow you to define a tensor with a given shape, but some strides set to zero so it implicitly repeats over those dimensions (like in broadcasting, but for a specific numer of times).

> It seems unnecessary to copy the same thing n times.

Depending on what you're doing with the tensor, you may actually require it to be explicitly repeated (so you can modify elements in place without affecting their repetitions). Otherwise we could ask upstream how difficult it would be to add something like `as_strided` to Theano. I guess it would require a lot of changes...

**Comment by skaae** [comment]
@benanne: I guess you can see it as a simple hack to create an encoder-decoder.
In the encoder/decoder you want to feed the last hidden state into the decoder. A simple way to do that is to use a slice layer to get the last hidden state and then repeat it `num_decoder_steps` times.

@f0k: No i havent profiled other options. I can try to do that later today.

**Comment by benanne** [comment]
@skaae, I figured it was something like that, but as @f0k pointed out this shouldn't really require anything more than adding an additional dimension with a stride of 0. I suppose we won't see `as_strided` in Theano anytime soon as optimizations may actually change the strides of arrays.

Ideally it would be possible to propagate shapes and strides through a Theano graph, then most of Lasagne would probably become obsolete :p

**Comment by nouiz** [comment]
as_strided, to be done correctly need a few changes in Theano:
1) Make an op and implement its gradient and opt to move it to the GPU
(easy for advanced user)
2) Make the output of the node as not able to work in place (not too hard
for core dev)
3) Optional, but recommandede before wide use, modify DebugMode to test
each node with such inputs to make sure that all node support it.

But we are pretty busy here iwh the summer school. If one of you want to
push on this, you can start 1 and ask again in 2-3 week that we do 2.

Fred

**Comment by f0k** [comment]
> Ideally it would be possible to propagate shapes and strides through a Theano graph, then most of Lasagne would probably become obsolete :p

Yeah, it also occurred to me that most of what `lasagne.layers` does is providing a wrapper of Theano Ops that can propagate shapes in Python code so we can create correctly-sized shared variables for parameters...

> as_strided, to be done correctly need a few changes in Theano:
> 1) Make an op and implement its gradient and opt to move it to the GPU (easy for advanced user) 
> 2) Make the output of the node as not able to work in place (not too hard for core dev)

That's less than I thought. But still too much for me right now, I wouldn't profit from this in the near future. Thanks a lot for the explanations, though!

**Comment by nouiz** [comment]
On Fri, Aug 7, 2015 at 9:33 AM, Sander Dieleman <notifications@github.com>
wrote:

> @skaae <https://github.com/skaae>, I figured it was something like that,
> but as @f0k <https://github.com/f0k> pointed out this shouldn't really
> require anything more than adding an additional dimension with a stride of
> 0. I suppose we won't see as_strided in Theano anytime soon as
> optimizations may actually change the strides of arrays.
>

I don't the problem you see with optimization. If all op handle correctly
stride of 0s, the only thing needed is the op and to disable inplace
optimization on that variable.

the step 3 would be good to do. Even if numpy have as_strided, not all
numpy function handle it correctly. This make it hard to rely on it...

If 3 end up being too hard, we could just put an attribute on Op that was
tested to support it and make a pass after the graph is optimizated to make
sure that only op that support it use the output of as_strided.

> Ideally it would be possible to propagate shapes and strides through a
> Theano graph, then most of Lasagne would probably become obsolete :p
>
There is @kastnerkyle that worked on something to help know the shapes when
building the graph. I don,t know if this interrest you. @kastnerkyle, can
you make an issue with the pointer to your code? That way other people can
see it before we get it in Theano.

For strides, the problem is that optimization can change that, so it would
be mostly an hint, strides is probably that, but nothing is sure. Shape is
asserted to be that (constant or a symbolic graph that is valid). But for
the strides, currently we don't have time for this. How important it is for
you to infer it? To my knowledge, only the meta optimizer could use it for
now. In a few days, we could make the needed subset op neede for that case
have its strides implemented, but we have more priority stuff to do now.

**Comment by f0k** [comment]
> But for the strides, currently we don't have time for this. How important it is for you to infer it? To my knowledge, only the meta optimizer could use it for now.

It's also that it could be used to optimize chains involving dimshuffles and gpu_contiguous() calls. I wouldn't give it high priority either, though.

**Comment by kastnerkyle** [comment]
Sure - I get shape information for symbolic variables using this function - https://github.com/dagbldr/dagbldr/blob/master/dagbldr/utils/utils.py#L213 and this one - https://github.com/dagbldr/dagbldr/blob/master/dagbldr/utils/utils.py#L284

Basically I keep an OrderedDict which has all shared variables, Theano random streams, and a minibatch of the data, labels, etc. Then every layer knows what the input data dimensionality is, and you can walk the fgraph to figure out all the other shapes. The only real issue is some strange interactions when trying to do this inside a step function in scan, but I think that can be overcome with a little more effort (I haven't *had* to just yet).

All that said - I don't really get why you need this repeat layer for encoder-decoder? I have done encoder decoder and encoder decoder with attention in my framework with shape inference but there are many other implementations that don't do shape inference and it works fine. Maybe this has to do with Lasagne internals?

**Comment by f0k** [comment]
> Sure - I get shape information for symbolic variables using [...]

Thanks! But `alt_shape_of_variables()` uses `theano.function()` -- it would be nicer to have a way that just works directly in Python, without any costly compilation in between. That's what we're replicating in Lasagne.

> All that said - I don't really get why you need this repeat layer for encoder-decoder?

We have to wait for @skaae to comment on that. I imagine it's because `theano.scan` needs correctly-shaped tensors in `outputs_info`, not just something that is broadcastable against the output shape of `fn`.

**Comment by skaae** [comment]
Yes its because we scan over the sequence lenght dimension. You could of course just write a decoder layer which didn't do that. 

E.g. a decoder layer which takes an num_decoder_steps argument.

**Comment by kastnerkyle** [comment]
You have to compile the function but it is not costly - FAST_COMPILE makes
this basically invisible. I call this hundreds of times in my code (think
several times per layer, per input to the layer) and the overall compile
time doesn't have a difference at all. As far as I know, you *have* to
compile a function unless you want to track shapes in the .name or .tag
(which is what I had before - not pleasant).

For outputs_info the initial hidden is supposed to be a projection from the
last hidden state (see
https://github.com/dagbldr/dagbldr/blob/master/dagbldr/nodes/layers.py#L518)
which should be an input to this "layer". If this is the case you should
have all the shapes needed in the inputs (assuming user has to define
hidden dim). If you init with zeros (which can work, but is worse) you can
do tensor.zeros_like(some_hidden_tensor_slice) assuming you are passing the
whole hidden sequence in - otherwise you can play other tricks. I really
recommend requiring previous hiddens as input - you need this for
bidirectional model with attention anyways.

You can definitely scan without needing that shape directly - that kind of
defeats the point of scan. If you have an implementation online I can take
look and use the right variable names, etc. to describe what I mean.

**Comment by kastnerkyle** [comment]
If you want to see my usage - check the tests
https://github.com/dagbldr/dagbldr/blob/master/dagbldr/tests/test_full_encdec.py#L36

I have not published examples yet but the tests are basically "examples
lite"

**Comment by f0k** [comment]
> You can definitely scan without needing that shape directly - that kind of defeats the point of scan. If you have an implementation online I can take look and use the right variable names, etc. to describe what I mean.

The problem I had in mind was when we have an initial hidden state vector (even if you provide previous hidden states as input whenever possible, at some point you *do* need an initial hidden state because the observed sequence starts and ends somewhere), we need to replicate it `batch_size` times to be able to pass it to scan. In our code, this happens via a dot product here: https://github.com/Lasagne/Lasagne/blob/master/lasagne/layers/recurrent.py#L357-L358
Just making it broadcastable across the batch dimension didn't work. (At least that's what Søren found.)

> As far as I know, you *have* to compile a function unless you want to track shapes in the .name or .tag (which is what I had before - not pleasant).

Theano has a `getOutputShape()` interface for Ops: https://github.com/Theano/Theano/blob/master/theano/tensor/nnet/conv.py#L254
It's just that it's not required for Ops to implement, that's why you need `theano.function`. If `FAST_COMPILE` does something clever avoiding compilation, it might be useful to tap into that, though.

**Comment by skaae** [comment]
@kastnerkyle: Given that we are in the same room we might discuss this in person :) I have a few other questions about Blocks / Fuel that you might be able to answer?

**Comment by kastnerkyle** [comment]
@skaae I can probably talk during the Fuel / Theano session this afternoon
about this.

Isn't batch_size implicitly stored in the input tensor? If so you could
just do tensor.ones_like(inputs[0][0]) to init the hiddens - the mask also
defines the batch size but it seems optional for you here.

For the else: clause I guess it is checking if it is a numpy array?

FAST_COMPILE just avoids all optimizations and (generally) just uses Python
code. So you are compiling a function but it is basically just a bunch of
Python Ops tied together.

**Comment by f0k** [comment]
> If so you could just do tensor.ones_like(inputs[0][0]) to init the hiddens

No, we don't want to initialize it to ones, but to `self.hid_init` replicated `batch_size` times, where `self.hid_init` is a shared variable that is optionally learned. And it seemed `scan` doesn't allow us to use broadcasting instead of explicit replication.

**Comment by kastnerkyle** [comment]
Talking specifically about this
https://github.com/Lasagne/Lasagne/blob/master/lasagne/layers/recurrent.py#L358

I don't see any reason you couldn't generate different set of ones then
broadcast against that.

**Comment by f0k** [comment]
> Talking specifically about this

I'm also talking specifically of this, but maybe you misunderstood what that line is doing. As the comment indicates, we're using a dot product against `T.ones` to replicate `self.hid_init` a given number of times. What we'd like to do instead would be broadcasting `self.hid_init`, without this explicit replication (via a dot product against a column vector of ones).

**Comment by skaae** [comment]
I don't know exactly where this settled but I'll profile 

* `T.stack`
* `T.tile`
* `T.repeat`

**Comment by mzhai2** [comment]
I'm using this for now:

```
